src/future/batch_semaphore.rs

```diff
@@ -0,0 +1,606 @@
+            }
+        }
+        drop(state);
+
+        // Releasing a semaphore is a yield point
+        crate::runtime::thread::switch();
+    }
+}
+
+// Safety: Semaphore is never actually passed across true threads, only across continuations. The
+// RefCell<_> type therefore can't be preempted mid-bookkeeping-operation.
+// TODO we shouldn't need to do this, but RefCell is not Send, and anything we put within a Semaphore
+// TODO needs to be Send.
+unsafe impl Send for BatchSemaphore {}
+unsafe impl Sync for BatchSemaphore {}
+
+impl Default for BatchSemaphore {
+    fn default() -> Self {
+        Self::new(Default::default(), Fairness::StrictlyFair)
+    }
+}
+
+/// The future that results from async calls to `acquire*`.
+/// Callers must `await` on this future to obtain the necessary permits.
+#[derive(Debug)]
+pub struct Acquire<'a> {
+    waiter: Arc<Waiter>,
+    semaphore: &'a BatchSemaphore,
+    completed: bool, // Has the future completed yet?
+}
+
+impl<'a> Acquire<'a> {
+    fn new(semaphore: &'a BatchSemaphore, num_permits: usize) -> Self {
+        let waiter = Arc::new(Waiter::new(num_permits));
+        Self {
+            waiter,
+            semaphore,
+            completed: false,
+        }
+    }
+}
+
+impl Future for Acquire<'_> {
+    type Output = Result<(), AcquireError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        assert!(!self.completed);
+        if self.waiter.has_permits.load(Ordering::SeqCst) {
+            assert!(!self.waiter.is_queued.load(Ordering::SeqCst));
+            self.completed = true;
+            trace!("Acquire::poll for waiter {:?} with permits", self.waiter);
+            Poll::Ready(Ok(()))
+        } else if self.semaphore.is_closed() {
+            assert!(!self.waiter.is_queued.load(Ordering::SeqCst));
+            self.completed = true;
+            trace!("Acquire::poll for waiter {:?} with closed", self.waiter);
+            Poll::Ready(Err(AcquireError::closed()))
+        } else if self.waiter.is_queued.load(Ordering::SeqCst) {
+            trace!("Acquire::poll for waiter {:?} already queued", self.waiter);
+
+            // Sanity check: there should be a waker, because the waiter is in
+            // the queue. Also true for unfair semaphores, which wake by ref.
+            assert!(self.waiter.waker.lock().unwrap().is_some());
+
+            // If the semaphore is fair, simply return. This corresponds to the
+            // future being polled after already having enqueued. No progress
+            // is made here, the semaphore will notify this waiter once ready.
+            if matches!(self.semaphore.state.borrow().fairness, Fairness::StrictlyFair) {
+                return Poll::Pending;
+            }
+
+            // Otherwise, try to acquire permits again. The result depends on
+            // which waiter(s) won the race after the `release` call.
+            let mut state = self.semaphore.state.borrow_mut();
```

Comment:
The code below is, apart from the `self.semaphore.remove_waiter(&self.waiter);` in the `Ok(())` branch, identical with the `else` branch below. Is it possible to condense the two pieces of code into one?